### DISCUSSION TITLE
Authentication layer, structure and microservice

### DISCUSSION THREAD

**Discussion by kachar** [podkrepi-bg/frontend#259] [General]
We need authentication layer to be used by the `graphql-gateway` in order to authenticate requests before sending them to their destination microservice.

It will be best if this layer also provides a frontend authentication mechanism/library for easier integration just like https://github.com/supabase/gotrue-js handles the frontend layer of https://github.com/netlify/gotrue

Reading material:

### Overview

- https://github.com/netlify/gotrue
- Demo https://gotruejs-playground.netlify.app/
- https://supabase.io/docs/guides/auth
- https://supabase.io/docs/guides/storage#add-security-rules
- https://github.com/supabase/supabase
- https://gqlgen.com/recipes/authentication/
- https://github.com/markbates/goth
- https://github.com/jrapoport/gothic

![image](https://user-images.githubusercontent.com/893608/121751871-af42a980-cb17-11eb-9e5a-b34cc44a0405.png)

### Supported providers

- Google
- GitHub
- GitLab
- Azure
- Facebook
- Bitbucket
- Twitter
- Apple

### Pros

- works with postgres out of the box
- can manage storage authentication and ACL https://supabase.io/blog/2021/03/30/supabase-storage
- works with Row Level Security (RLS) via policies
- has client libraries for js, c#, go
- has rest interface

### Overview

- https://www.keycloak.org/
- https://github.com/keycloak/keycloak
- https://www.keycloak.org/docs/latest/server_admin/

![image](https://user-images.githubusercontent.com/893608/121748666-080f4380-cb12-11eb-9fea-ce8e803416ca.png)

### Supported providers

https://www.keycloak.org/docs/latest/server_admin/#social-identity-providers

- Google
- Facebook
- Twitter
- GitHub
- LinkedIn
- Microsoft
- Stack Overflow

### Cons

- separate application
- integrates with redirects

### Overview

- https://next-auth.js.org/
- https://github.com/nextauthjs/next-auth
- Demo https://next-auth-example.vercel.app/

### Supported providers

- Apple
- Auth0
- Credentials
- Discord
- Email
- Facebook
- GitHub
- Google
- Instagram
- LinkedIn
- Okta
- Slack
- Twitter
- VK
- Atlassian, Azure Active Directory B2C, Basecamp, Battle.net, Box, Bungie, Amazon Cognito, Coinbase, Dropbox, EVE Online, FACEIT, Foursquare, FusionAuth, GitLab, IdentityServer4, Kakao, LINE, Mailchimp, Mail.ru, Medium, Netlify, Osso, Reddit, Salesforce, Spotify, Strava, Twitch, WordPress.com, WorkOS, Yandex, Zoho, Zoom, 

### Pros

- nice frontend integration

### Cons

- no backend authentication between microservices

**Comment by cefothe** [comment]
During the analyst of Keycloak, the redirects are not required. Because the Keycloak is a separate application we can scale it out of the box and give us the flexibility to extend the functionally that provides to us.

**Comment by cefothe** [comment]
Based on GitHub, the last release of GoTrue is last year which is a red flag in my head. But https://github.com/jrapoport/gothic looks pretty OK. So we can try that, the pros: written in Go, active community but the last release was March.

**Reply by cefothe** [reply]
@kachar What you think about that?

**Reply by kachar** [reply]
It looks more advanced than the base netlify version, which is good

I like that this project supports more providers via goth https://github.com/markbates/goth

> Gothic uses goth for external oauth providers now. Now we support everything that [goth](https://github.com/markbates/goth) supports.


---

I'm not sure how it compares to the Supabase fork of gotrue, which is still activelly maintained at https://github.com/supabase/gotrue

The latest release has been 19 days ago

As it's part of supabase core products I'm pretty sure they'll maintain it for the community on the long run.

**Reply by cefothe** [reply]
We can try both projects and check them how they work ... But almost 100 % supabase fork will more easy at the begging.

**Comment by cefothe** [comment]
We can check https://medevel.com/iam-systems-10-identity/ for more ideas and tools that we can use. We can try to use something like GoThic or KeyClock both of them are definitely good choices that we can use.

**Comment by cefothe** [comment]
I am not sure if the free version of https://supabase.io/pricing will make sense for us.

**Reply by kachar** [reply]
Even if the free version makes sense for use we better deploy the stack on premisse to avoid data leaks, to guarantee transparancy and to evolve the stack naturally with additional internal components.

One example might be to provide campaigns module to access the database locally without going to Supabase infrastructure

**Reply by cefothe** [reply]
Totally agree with your statement, so we can exclude supabase from options.

**Reply by kachar** [reply]
Nevertheless supabase provides a bundle of opensource tools that work together nicely, so even if we exclude their SaaS solution I think we can keep the OSS solution.

**Reply by cefothe** [reply]
Can you share the links to those tools? Because the configuration can be difficult as well ...

**Reply by cefothe** [reply]
@kachar I get what you mean. Sorry
https://supabase.io/assets/images/supabase-architecture-9050a7317e9ec7efb7807f5194122e48.png

**Reply by kachar** [reply]
The core repo is at https://github.com/supabase/supabase

You can see the docker compose setup at https://github.com/supabase/supabase/blob/master/docker/docker-compose.yml

It includes:

- kong (https://github.com/Kong/kong) + config (https://github.com/supabase/supabase/blob/master/docker/dockerfiles/kong/kong.yml)
- postgre + plugins (https://github.com/supabase/postgres)
- postgrest (https://postgrest.org/en/stable/) + js lib (https://github.com/supabase/postgrest-js)
- storage server + rest api (https://github.com/supabase/storage-api)
- realtime server (https://github.com/supabase/realtime)
- gotrue authentication (https://github.com/supabase/gotrue)

**Reply by cefothe** [reply]
We are not going to need the entire setup in the beginning. Probably only:
1. Kong
2. Storage server + REST API (not sure if we will have enough money to use s3)  
3. GoTrue (they fork, not the original one). @kachar Do you have some information about which standards they support, I only can see OAuth...

**Reply by kachar** [reply]
About the S3 storage - there are 3rd party oss packages that support this storage protocol just like https://github.com/minio/minio
In this way we avoid AWS storage fees (which are not that great anyway, like $1 per GB per month or less [[Source]](https://aws.amazon.com/s3/pricing/), [AWS estimate](https://calculator.aws/#/estimate?id=409bb7ad5b82b08a237f415d7160d45c0b1881d4))

About the supported providers - as I see they are limited to `azure`, `bitbucket`, `github`, `gitlab`, `facebook`, `twitter`, `apple` and `google`

**Reply by cefothe** [reply]
Storage can be a separate discussion because we need to be highly available and fail tolerance. If we can have encryption and versioning that will be perfect. But anyway we can reuse as much as we can in order to minimize the work, so we can just create a different adapter. About s3 comes with a lot of those things out of the box, we really need to think about on-premise that can provide the same functionality.

 Supported provides are pretty ok ...

**Reply by cefothe** [reply]
Some s3 alternatives https://geekflare.com/self-hosted-s3/.

**Reply by kachar** [reply]
Let's split that in separate storage related discussion, the link seems promising. Would you start one?

**Reply by preslavgerchev** [reply]
If we go for `supabase` as an auth layer, does that mean we should strongly consider `Kong` as a [gateway](https://github.com/podkrepi-bg/frontend/discussions/260)? I would imagine it makes sense to simply extend the existing `Kong` deployment in that case and get both? 👀

**Reply by cefothe** [reply]
@preslavgerchev It looks like the most simple way is to start with supabase as an auth layer and Kong as a gateway. We are searching for something that will give us a more easy and fast way to bootstrap the main part of the project. And defiantly the supabase is a promising solution. 

About the second point, I can't get what you mean with the existing `Kong` deployment, is that mean we already have `Kong` in the project.

**Reply by preslavgerchev** [reply]
It was more of a note than anything that `Kong` is [deployed](https://github.com/supabase/supabase/blob/master/docker/dockerfiles/kong/Dockerfile) as part of the `supabase` stack. In theory we could reuse this deployment for API gateway purposes as well, although I am not sure how feasible or desirable that is 😄
